File: static/sorting.js
const Sorting = {
    autoSort(players) {
        const shuffled = [...players].sort(() => Math.random() - 0.5);
        const tanks = shuffled.filter(p => p.role === "Tank");
        const healers = shuffled.filter(p => p.role === "Healer");
        const melee = shuffled.filter(p => p.role === "Melee");
        const ranged = shuffled.filter(p => p.role === "Ranged");
        const groups = [];
        while (tanks.length && healers.length && (melee.length + ranged.length) >= 3) {
            const group = [];
            const tank = tanks.shift(); group.push(tank);
            const healer = this._pullUtil(healers, !hasLust(tank.wow_class), !hasBrez(tank.wow_class));
            group.push(healer);
            let gl = hasLust(tank.wow_class) || hasLust(healer.wow_class);
            let gb = hasBrez(tank.wow_class) || hasBrez(healer.wow_class);
            for (let s = 0; s < 3; s++) {
                const pref = s === 0 ? true : s === 1 ? false : null;
                const dps = this._pullDPS(melee, ranged, !gl, !gb, pref);
                if (dps) { group.push(dps); gl = gl || hasLust(dps.wow_class); gb = gb || hasBrez(dps.wow_class); }
            }
            groups.push(group);
        }
        return { groups, bench: [...tanks, ...healers, ...melee, ...ranged] };
    },
    _pullUtil(pool, nL, nB) { const i = this._findUtil(pool, nL, nB); return i > -1 ? pool.splice(i, 1)[0] : pool.shift(); },
    _findUtil(pool, nL, nB) {
        for (let i = 0; i < pool.length; i++) { if (nL && hasLust(pool[i].wow_class) && nB && hasBrez(pool[i].wow_class)) return i; }
        for (let i = 0; i < pool.length; i++) { if ((nL && hasLust(pool[i].wow_class)) || (nB && hasBrez(pool[i].wow_class))) return i; }
        return -1;
    },
    _pullDPS(melee, ranged, nL, nB, pref) {
        let pri, sec;
        if (pref === true) { pri = melee; sec = ranged; } else if (pref === false) { pri = ranged; sec = melee; }
        else { pri = melee.length >= ranged.length ? melee : ranged; sec = melee.length >= ranged.length ? ranged : melee; }
        if (nL || nB) { for (const p of [pri, sec]) { const i = this._findUtil(p, nL, nB); if (i > -1) return p.splice(i, 1)[0]; } }
        if (pri.length) return pri.shift(); if (sec.length) return sec.shift(); return null;
    },
};

File: New folder (4)/app/services/sorting.py
"""
NightOwls Smart Group Sorting — V4
Builds 5-man M+ groups: 1 Tank, 1 Healer, 3 DPS
Prioritizes Lust + Brez per group, balances melee/ranged, shuffles for fairness.
"""
import random
from app.models.schemas import has_lust, has_brez


def auto_sort(players: list[dict]) -> dict:
    players = [p.copy() for p in players]
    random.shuffle(players)

    tanks = [p for p in players if p["role"] == "Tank"]
    healers = [p for p in players if p["role"] == "Healer"]
    melee = [p for p in players if p["role"] == "Melee"]
    ranged = [p for p in players if p["role"] == "Ranged"]

    groups = []

    while tanks and healers and (len(melee) + len(ranged)) >= 3:
        group = []
        tank = tanks.pop(0)
        group.append(tank)

        healer = _pull_utility(healers, not has_lust(tank["wow_class"]), not has_brez(tank["wow_class"]))
        group.append(healer)

        g_lust = has_lust(tank["wow_class"]) or has_lust(healer["wow_class"])
        g_brez = has_brez(tank["wow_class"]) or has_brez(healer["wow_class"])

        for slot in range(3):
            prefer = True if slot == 0 else (False if slot == 1 else None)
            dps = _pull_dps(melee, ranged, not g_lust, not g_brez, prefer)
            if dps:
                group.append(dps)
                g_lust = g_lust or has_lust(dps["wow_class"])
                g_brez = g_brez or has_brez(dps["wow_class"])

        groups.append(group)

    bench = tanks + healers + melee + ranged
    return {"groups": groups, "bench": bench}


def _pull_utility(pool, need_lust, need_brez):
    idx = _find_utility(pool, need_lust, need_brez)
    return pool.pop(idx) if idx > -1 else pool.pop(0)


def _find_utility(pool, need_lust, need_brez):
    for i, p in enumerate(pool):
        if need_lust and has_lust(p["wow_class"]) and need_brez and has_brez(p["wow_class"]):
            return i
    for i, p in enumerate(pool):
        if (need_lust and has_lust(p["wow_class"])) or (need_brez and has_brez(p["wow_class"])):
            return i
    return -1


def _pull_dps(melee, ranged, need_lust, need_brez, prefer_melee):
    if prefer_melee is True:
        pri, sec = melee, ranged
    elif prefer_melee is False:
        pri, sec = ranged, melee
    else:
        pri, sec = (melee, ranged) if len(melee) >= len(ranged) else (ranged, melee)

    if need_lust or need_brez:
        for pool in [pri, sec]:
            idx = _find_utility(pool, need_lust, need_brez)
            if idx > -1:
                return pool.pop(idx)

    if pri:
        return pri.pop(0)
    if sec:
        return sec.pop(0)
    return None
